### PR TITLE
[mini] Eliminate some sources of quadratic behavior. Improves on #7809.

### PR DESCRIPTION
**Pull request by kumpera**
This reduces compilation times for #7809 by 20% on OSX amd64.
It still takes 46 seconds on a good keyboard MBP.

### CODE REVIEW

File: mono/mini/liveness.c [hunk 1]
```diff
@@ -37,6 +37,62 @@
...
+
+static void
+mono_ptrset_add (MonoPtrSet *set, gpointer val)
+{
+	//switch to hashtable
+	if (set->capacity == INLINE_SIZE) {
+		GHashTable *tmp = g_hash_table_new (NULL, NULL);
+		for (int i = 0; i < INLINE_SIZE; ++i)
+			g_hash_table_insert (tmp, set->data [i], set->data [i]);
+		set->data [0] = tmp;
```
Comment (vargaz) [inline mono/mini/liveness.c]: Would be nicer to store the hash as an additional field.

File: mono/mini/liveness.c [hunk 2]
```diff
@@ -37,6 +37,62 @@
 static void mono_analyze_liveness2 (MonoCompile *cfg);
 #endif
 
+
+#define INLINE_SIZE 16
```
Comment (lewurm) [inline mono/mini/liveness.c]: did you play around with other values?

File: mono/mini/method-to-ir.c [hunk 3]
```diff
@@ -657,52 +657,78 @@ mono_link_bblock (MonoCompile *cfg, MonoBasicBlock *from, MonoBasicBlock* to)
...
+static void
+mark_bb_in_region (MonoCompile *cfg, guint region, uint32_t start, uint32_t end)
+{
+	MonoBasicBlock *bb = cfg->cil_offset_to_bb [start];
+
+	//start must exist in cil_offset_to_bb as those are il offsets used by EH which should have GET_BBLOCK early.
+	g_assert (bb);
+
+	if (cfg->verbose_level)
+		printf ("FIRST BB for %d is BB_%d\n", start, bb->block_num);
```
Comment (lewurm) [inline mono/mini/method-to-ir.c]: `s/printf/g_print`

File: mono/mini/method-to-ir.c [hunk 4]
```diff
@@ -657,52 +657,78 @@ mono_link_bblock (MonoCompile *cfg, MonoBasicBlock *from, MonoBasicBlock* to)
...
 	}
-	for (i = 0; i < header->num_clauses; ++i) {
-		clause = &header->clauses [i];
 
-		if (MONO_OFFSET_IN_CLAUSE (clause, offset))
-			return ((i + 1) << 8) | clause->flags;
+	if (cfg->verbose_level > 2) {
+		MonoBasicBlock *bb;
+		for (bb = cfg->bb_entry; bb; bb = bb->next_bb)
+			printf ("REGION BB%d IL_%04x ID_%08X\n", bb->block_num, bb->real_offset, bb->region);
```
Comment (lewurm) [inline mono/mini/method-to-ir.c]: `s/printf/g_print`

File: mono/mini/liveness.c (hunk 2 above)
Comment (kumpera) [reply mono/mini/liveness.c]: Yes, I measured 8 and 32 with Roslyn and 32 was the best one.

File: mono/mini/liveness.c (hunk 1 above)
Comment (kumpera) [reply mono/mini/liveness.c]: That would waste memory thou

**Comment by kumpera** [comment]
Did s/printf/g_print

File: mono/mini/liveness.c (hunk 2 above)
Comment (lewurm) [reply mono/mini/liveness.c]: why 16 then? 😄 or did you mean 16 instead of 32?

**Comment by kumpera** [comment]
@monojenkins build failed

**Comment by vargaz** [comment]
We map printf to g_print in mini.h, so it doesn't matter right now which one is used btw.

**Comment by lewurm** [comment]
@vargaz only on android:
https://github.com/mono/mono/blob/626ab1064cf53df05eaaa1828916c421b08fa56e/mono/mini/mini.h#L127-L130

File: mono/mini/mini-amd64.c [hunk 5]
```diff
@@ -2460,7 +2463,7 @@ mono_arch_emit_setret (MonoCompile *cfg, MonoMethod *method, MonoInst *val)
 	        x86_branch (code, cond, cfg->native_code + ins->inst_true_bb->native_offset, sign); \
         } else { \
 	        mono_add_patch_info (cfg, code - cfg->native_code, MONO_PATCH_INFO_BB, ins->inst_true_bb); \
-	        if ((cfg->opt & MONO_OPT_BRANCH) && \
+	        if (optimize_branch_pred && \
```
Comment (jaykrell) [inline mono/mini/mini-amd64.c]: Shouldn't the conditions be and'ed?

File: mono/mini/liveness.c (hunk 2 above)
Comment (kumpera) [reply mono/mini/liveness.c]: What I meant was that I tried 8, 16 and 32. The winner was 16.

**Comment by kumpera** [comment]
@monojenkins build failed

**Pull request merged**